File: Data Structures & Algorithms/subsets-ii/submission-10.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]: 
        nums.sort()
        res = [] 

        def dfs(start, part): 
            res.append(part.copy()) 

            for i in range(start, len(nums)): 
                if i > start and nums[i] == nums[i - 1]: 
                    continue 

                part.append(nums[i]) 
                dfs(i + 1, part) 
                part.pop() 

        dfs(0, []) 
        return res

File: Data Structures & Algorithms/palindrome-partitioning/submission-10.js
class Solution {
    /**
     * @param {string} s
     * @return {string[][]}
     */
    partition(s) { 
        const res = [] 
        this.dfs(0, s, [], res) 
        return res
    } 

    dfs(start, s, part, res) {
        if ( start === s.length ) {
            res.push([...part]) 
            return
        } 

        for (let i = start; i < s.length; i ++) {
            const partition = s.slice(start, i + 1) 
            if (this.isPali(partition)) {
                part.push(partition) 
                this.dfs(i + 1, s, part, res) 
                part.pop()
            }
        }
    } 

    isPali(string) {
        let l = 0 
        let r = string.length - 1 
        while(l < r) {
            if (string[l] !== string[r]) {
                return false
            } 
            l ++ 
            r --
        } 

        return true
    }
}
